src/order/upset.lean

```diff
@@ -0,0 +1,290 @@
+/-
+Copyright (c) 2022 Yaël Dillies, Sara Rousta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Sara Rousta
+-/
+import data.set.lattice
+import data.set_like.basic
+
+/-!
+# Up-sets and down-sets
+
+This file defines upper and lower sets in an order.
+
+## Main declarations
+
+* `is_upset`: Predicate for a set to be an upper set.
+* `is_downset`: Predicate for a set to be a lower set.
+* `up_set`: The type of upper sets.
+* `down_set`: The type of lower sets.
+-/
+
+open set
+
+variables {ι : Sort*} {κ : ι → Sort*} {α : Type*}
+
+/-! ### Unbundled upper/lower sets -/
+
+section unbundled
+variables [has_le α] {s t : set α}
+
+/-- An upper set in an order `α` is a set such that any element greater than one of its members is
+also a member. -/
+def is_upset (s : set α) : Prop := ∀ ⦃a b⦄, a ≤ b → a ∈ s → b ∈ s
+
+/-- A lower set in an order `α` is a set such that any element less than one of its members is also
+a member. -/
+def is_downset (s : set α) : Prop := ∀ ⦃a b⦄, b ≤ a → a ∈ s → b ∈ s
```

Comment:
Can you add the `order_dual` constructions to turn `is_downset s` into `is_upset (order_dual.of_dual \inv' s)` or whatever the real statement is?